Add support for edit history

preetmishra opened the pull request:
This PR builds up on @shreyamalviya's #354 to show edit history for a message. :+1:

The popup can be triggered by pressing  `d`  from the `MsgInfoView` for any message (we could discuss other key options and whether it would be better to show the popup only for messages that are edited).

For **v1**, the popup only shows the content for every snapshot in the 'message_history' (not the diff). 

I have split the last two commits for making it easier to review. We could squash the two if necessary.

Fixes #135 partially.

preetmishra commented:
I'll be adding the tests once we reach a consensus on the design and the implementation details.

preetmishra commented:
Updated with an improved theme for retro mode.

preetmishra commented on zulipterminal/core.py:
```diff
@@ -124,6 +125,10 @@ def show_stream_info(self, color: str, name: str, desc: str) -> None:
     def popup_with_message(self, text: str, width: int, height: int) -> None:
         self.show_pop_up(NoticeView(self, text, width, height), "NOTICE")
 
+    def show_edit_history(self, msg: Any) -> None:
```
Thanks for pointing it out.

preetmishra commented:
@sumanthvrao Thanks for the review.

I have fixed the bug that you reported; you should be able to the history as soon as you make an edit (it was an issue with `edit_history`). In addition to that, I have added the prefixes like the web app (e.g `Edited by XYZ`).

Re double edit, I am not sure what do you want to suggest by 'different'. Could you elaborate a bit?

I think that precise timestamps would be useful for distinguishing messages accurately (consider a message that has been edited thrice within a minute).

We could discuss whether it would be better to not show `EditHistoryView` for messages that have not been edited. It would be straightforward to implement with `edit_history` but `edit_history` isn't reliable at the moment (see [#issues>edit history isn't dynamic](https://chat.zulip.org/#narrow/stream/9-issues/topic/edit.20history.20isn't.20dynamic/near/882811)). 

Also, `edit_history` is likely to be changed to a boolean for checking if the message has been edited or not (see [#backend>edit_history](https://chat.zulip.org/#narrow/stream/3-backend/topic/edit_history/near/879944)), so the issue might get fixed along the way.

sumanthvrao commented:
@preetmishra Tried the updated PR out, it's a nice improvement 👌 . The edit-from-zt bug is gone now.

By double edit I mean you edit a message twice and the second edit wasn't seen. But I tried this after your rebase and I'm no longer able to reproduce it. I think it was coupled with the edit-from-ZT bug which you fixed.

Re `EdithistoryView` for non-edited, Sure we can start a discussion on the stream for this. Ah! boolean value would simplify some logic we have. Nice!

I think you can proceed to add tests for the PR.

preetmishra commented:
@sumanthvrao Thanks for taking another look. I am glad that the 'double edit' and 'edit-from-ZT' bug are sorted for you. :+1: 

I updated the PR with the tests. It covers every aspect except the components related to `bulid_pile` in `EditHistoryView`. I am not sure with the level of detail that I should cover. Could you share some pointers on how to go about it?

preetmishra commented:
@sumanthvrao Thanks for taking another look. I am glad that the 'double edit' and 'edit-from-zt' bug are sorted for you. :+1: 

I updated the PR with the tests. It covers every aspect except the ones that related to `bulid_pile` in `EditHistoryView`. I am not sure with the level of detail that I should cover. Could you share some pointers on how to go about it?

sumanthvrao commented on README.md:
```diff
@@ -155,6 +155,7 @@ source ~/.zshenv
 | Add/remove thumbs-up reaction to the current message  | <kbd>+</kbd>                                  |
```
@preetmishra It maybe simpler here to split the commit into 
a) README/keys - which adds the key (without linking it) (`git log --grep='README/keys'`)
b) EditHistory View  - and test for it.
c) Finally, link the keypress function - and tests for it.

Splitting the commits in such a fashion makes reviewing easier and commit much more coherent.

preetmishra commented:
@sumanthvrao Thanks for the clarification and the suggestion. :+1: 

I have split the second commit as per your suggestion in https://github.com/zulip/zulip-terminal/pull/663#discussion_r430823497. In addition to that, I have split the last commit into three separate commits for clarity.

Hopefully, this should be easier to review and comprehend.

preetmishra commented on tests/conftest.py:
```diff
@@ -312,6 +312,45 @@ def messages_successful_response() -> Dict[str, Any]:
+    """
+    return request.param
+
+
```
The 'message history' is also needed in `TestEditHistoryView` which is why I went ahead with a fixture in `conftest.py`.

preetmishra commented on zulipterminal/ui_tools/views.py:
```diff
@@ -925,3 +928,19 @@ def __init__(self, controller: Any, msg: Message) -> None:
+        self.msg = msg
+        self.width = 64
+        self.height = 0
+        super().__init__(controller, [], 'MSG_INFO')
```
The `MSG_INFO` makes me wonder if it'd be a better idea to rename the `self.command` variable to something like `self.exit_popup_command`.

sumanthvrao commented on zulipterminal/ui_tools/views.py:
```diff
@@ -967,27 +968,45 @@ def __init__(self, controller: Any, msg: Any) -> None:
 
         if 'prev_topic' in snapshot and topic != snapshot['prev_topic']:
             topic += ' (updated from {})'.format(snapshot['prev_topic'])
+            if 'prev_content' in snapshot:
```
We seem to show the fact that the topic has been updated, in both the title and the body here. We show it in `updated from {}` (Title) and show who it was updated by as well. Do you think sticking with `Edited by X` might be simpler here, even for a topic update? Might simplify things a bit. Might be a discussion to have on the stream.

sumanthvrao commented on zulipterminal/ui_tools/views.py:
```diff
@@ -967,27 +968,45 @@ def __init__(self, controller: Any, msg: Any) -> None:
+        if self.msg['sender_id'] == user_id:
+            return self.msg['sender_full_name']
+        else:
+            for user in self.controller.view.users:
```
While this does take `O(n)` in the worst case, it does if for every `edit` in edit-history. I was brainstorming something faster and maybe we can use a combination of `model.user_id_email_dict` and `model.user_dict` to get the name faster. Worthwhile to add a `FIXME` in case we decide to go with this for now.

preetmishra replied:
Thanks so much, the proposed strategy is faster than the `O(n)` approach.

preetmishra commented on README.md:
```diff
@@ -155,6 +155,7 @@ source ~/.zshenv
 | Add/remove thumbs-up reaction to the current message  | <kbd>+</kbd>                                  |
 | Add/remove star status of the current message         | <kbd>*</kbd>                                  |
```
As `e` _can_ be changed in the future, I have added `EDIT_HISTORY` to the commit title instead.

preetmishra commented on tests/conftest.py:
```diff
@@ -312,6 +312,45 @@ def messages_successful_response() -> Dict[str, Any]:
+            'content': 'Hello!',
+            'rendered_content': '<p>Hello!</p>',
+            'timestamp': 1530129122,
+            'topic': 'party at my houz',
```
Except the 'HTML' bit, everything else is required. I have added it here as it'd be required when we add the diff feature later.

preetmishra commented on zulipterminal/ui_tools/views.py:
```diff
@@ -904,6 +904,9 @@ def __init__(self, controller: Any, msg: Message) -> None:
             ('Date & Time', time.ctime(msg['timestamp'])[:-5]),
             ('Sender', msg['sender_full_name']),
             ('Sender\'s Email ID', msg['sender_email']),
+            ('Edit History',
```
Yes, however, the current implementation adds quotes ('') to the keys that are returned from `keys_for_command`. The quotes make it easier to differentiate the keys from the rest of the text.

preetmishra commented on tests/ui/test_ui_tools.py:
```diff
@@ -1222,6 +1222,46 @@ def test_keypress_navigation(self, mocker,
         super_keypress.assert_called_once_with(size, expected_key)
 
 
+class TestEditHistoryView:
```
@neiljp mentioned that it'd be better to move all the popup views and their tests altogether. 

Would it be better to hold this off until then?

sumanthvrao replied:
Sure. Makes sense.

preetmishra commented on zulipterminal/model.py:
```diff
@@ -356,6 +356,16 @@ def get_messages(self, *,
+        response = self.client.get_message_history(message_id=message['id'])
+        if response['result'] == 'success':
+            message['message_history'] = response['message_history']
+            return ''
```
The other `get_*` methods in `model.py` use a similar approach. Could you point out where this would the proposed tuple be helpful?

preetmishra commented on zulipterminal/ui_tools/views.py:
```diff
@@ -940,8 +940,53 @@ def __init__(self, controller: Any, msg: Any) -> None:
+        self.widgets = []
+        # FIXME: Limit calls by checking if the message has been edited using
+        # 'edit_history' boolean (requires server change).
+        assert self.controller.model.get_message_history(self.msg) == ''
```
It is likely to fail when ZT has trouble connecting to the server. In that case, it'll raise an AssertionError.

preetmishra commented on zulipterminal/config/keys.py:
```diff
@@ -205,6 +205,12 @@
+    ('EDIT_HISTORY', {
+        'keys': {'e'},
+        'help_text': 'View edit history from message information box',
+        'excluded_from_random_tips': True,
```
Opened [Help tips](https://chat.zulip.org/#narrow/stream/206-zulip-terminal/topic/Help.20tips) on CZO.

preetmishra commented:
@sumanthvrao Thanks for the review and the suggestions.

I have:
* added the test for `build_pile`,
* removed the redundant 'Topic Updated' excerpt from the author prefix to stick with `Edited by X`,
* replaced the `O(n)` logic with your proposed method
* and made other amendments as per your suggestions.

neiljp commented:
@preetmishra This is my first pass at looking at manually testing this after your updates, and it seems pretty good :+1: 

Thoughts from looking at a few messages with histories:
* If there is the original and updated topics in every case, it seems unnecessary to have the '(updated from <old topic name>)', and it actually looks a bit confusing if the topics have parentheses too!
* To emphasize what changed in each version, how about having 'Content'/'Topic'/'Content & Topic' before 'edited by <user>'?
* From the edit menu, `e` and `esc` behave the same (back to info menu), but `i` exits both levels of menu?

preetmishra commented:
@neiljp Thanks for the review and the suggestions. I have removed the `(updated from {})` bit and extended the author prefix to show 'Content'/'Topic'/'Content & Topic' before 'edited by' or 'Edited but no changes made' accordingly.

In addition to that, I have restructured the 4th commit (see in-line comment), extended the commit messages and tidied up the code for clarity.

Re keys, it's been taken from the original PR so I do not have strong opinions. I am not sure about the other keys but entering and exiting with `EDIT_HISTORY` key feels intuitive. 

Note: This update renames `build_pile` to `edit_block` (this might come handy in understanding previous discussions where we refer to it as `build_pile`).

preetmishra commented on zulipterminal/ui_tools/views.py:
```diff
@@ -953,7 +953,10 @@ def __init__(self, controller: Any, msg: Message, title: str) -> None:
+                  ('Sender\'s Email ID', msg['sender_email']),
+                  ('Edit History', 'Press {} to view'.format(
+                      ', '.join(map(repr, keys_for_command('EDIT_HISTORY'))))),
+                  ]),
```
Previously, this label extension bit was in the 3rd/'Introduce EditHistoryView...' commit.

preetmishra commented:
Updated to resolve conflicts.

preetmishra commented:
Updated to resolve conflicts.

preetmishra commented:
Updated to resolve conflicts.

preetmishra commented on zulipterminal/config/keys.py:
```diff
@@ -210,6 +210,12 @@
         'help_text': 'View message information',
         'key_category': 'msg_actions',
     }),
+    ('EDIT_HISTORY', {
```
Yes, this should be merged with other commits.

preetmishra commented on tests/ui/test_ui_tools.py:
```diff
@@ -1353,6 +1353,51 @@ def test_feature_level_content(self, mocker, zulip_version):
         )
 
 
+class TestEditHistoryView:
```
Agreed; we discussed moving the pop-up code and tests a while ago. I have moved the test for `EditHistoryView` into a new file.

I have only rebased and added tests in this particular commit, which I have now mentioned in the commit text. However, the subsequent commit was significantly different, so I amended it appropriately. (We discussed something similar for the view message in the browser PR.)

Thanks, fixed!

preetmishra commented on zulipterminal/model.py:
```diff
@@ -375,6 +375,16 @@ def get_messages(self, *,
         display_error_if_present(response, self.controller)
         return response['msg']
 
+    def get_message_history(self, message: Message) -> str:
```
Ah, yes. This was prior to the discussion we had re "fetch vs get". Thanks!

Yes, we need to fetch it each time to get the message history. Given that we would need to fetch it each time to see if the message history was updated (via the same endpoint), I don't think caching would be beneficial.

I have added a new `get_message_history()` along with `_fetch_message_history()` to handle this differently.

preetmishra commented on zulipterminal/ui_tools/views.py:
```diff
@@ -1133,8 +1133,53 @@ def __init__(self, controller: Any, msg: Message,
+                state = (
+                    'Original' if index == 0 else
+                    'Current' if index == len(message_history) - 1 else ''
+                )
```
Fixed! Also, renamed the variable to `tag` itself.

preetmishra commented on zulipterminal/ui_tools/views.py:
```diff
@@ -1180,6 +1185,11 @@ def edit_block(self, snapshot: Dict[str, Any], state: str) -> Any:
+    def get_author(self, user_id: int) -> str:
+        user_email = self.controller.model.user_id_email_dict[user_id]
+        full_name = self.controller.model.user_dict[user_email]['full_name']
+        return full_name
```
I have moved and extracted it as a separate commit.

preetmishra commented:
@neiljp Thanks for the review. In the latest update, along with the things that I have replied to in-line, I have:
* Polished the history fetch and extracted `user_name_from_id`.
* Improved the tests.
* Amended `MsgInfoView` to only show `EditHistoryView` for edited messages.
* Included an extra commit to amend the `Date & Time` label in `MsgInfoView`.

Re edit summary in `MsgInfoView`, that would be rather costly given that it'd require fetching `message_history` every time the pop-up is opened (even when the user doesn't want to see the message history).

preetmishra commented on zulipterminal/model.py:
```diff
@@ -387,6 +387,29 @@ def get_messages(self, *,
+        """
+        Returns message history for a message.
+        """
+        assert self._fetch_message_history(message_id=message_id) == ''
```
Yes, it would. This should be fixed with `fetch_message_history()` returning an empty list now for failed request.

preetmishra replied:
This was to assert against any implementation loophole. However, as you mention, it would have failed the assert and update the footer when the server was down as well.

This should be fixed now with `fetch_message_history()` returning an empty list for failed request.

preetmishra replied:
This was to assert against any implementation loophole. However, as you mention, it would have failed the assert and update the footer when the server was down as well.

This should've been fixed now with `fetch_message_history()` returning an empty list for failed request.

preetmishra commented on tests/ui_tools/test_popups.py:
```diff
@@ -0,0 +1,71 @@
+        self.message = {'id': 1}
+        self.edit_history_view = EditHistoryView(
+            controller=self.controller,
+            message=self.message,
```
Yes, but this seems like a better way to do it given that the tests do not care about anything else than the `id`. Moreover, `message_fixture` appended extra test cases unnecessarily.

preetmishra commented on zulipterminal/config/themes.py:
```diff
@@ -222,6 +234,14 @@
+        ('edit_author',    'yellow',          'black',
+         None,             YELLOW,            BLACK),
+        ('edit_tag',       'light blue',      'black',
+         None,             LIGHTBLUE,         BLACK),
```
Yes. I primarily adjusted the `edit_tag` style in other themes to fit better as well.

preetmishra replied:
Yes, it seems to personally. I have also adjusted the `edit_tag` style in other themes.

preetmishra replied:
Agreed. In the latest update, I have swapped `tag` with `time` and slightly adjusted the themes to improve this.

preetmishra commented on tests/ui/test_ui_tools.py:
```diff
@@ -1454,6 +1454,14 @@ class TestMsgInfoView:
+            'edited_messages': set((537286, 537287, 537288)),
+        }
+        self.controller.model.initial_data = {
+            'realm_allow_edit_history': True,
```
Since this and the patch above it are required for other tests in `TestMsgInfoView` as well, I only patched them once here. Though, I have simplified these in the latest update.

I am not sure if we can overwrite a fixture patch (this particular fixture, `mock_external_classes`, is an `autouse` fixture which is used everywhere in the class).

Besides, the `edit_history` test simply cares about the `show_edit_history_label` value (and how that value is set [with `and` between booleans] is trivial in comparison).

preetmishra commented on tests/ui/test_ui_tools.py:
```diff
@@ -1454,6 +1454,14 @@ class TestMsgInfoView:
+        self.controller.model.index = {
+            # These message IDs are from message_fixture. Patch
+            # show_edit_history_label separately wherever required.
+            'edited_messages': set((537286, 537287, 537288)),
```
See https://github.com/zulip/zulip-terminal/pull/663#discussion_r472314071.

preetmishra commented on zulipterminal/ui_tools/views.py:
```diff
@@ -1072,7 +1072,8 @@ def __init__(self, controller: Any, msg: Message, title: str,
+            ('', [('Date & Time (Original)',
+                   time.ctime(msg['timestamp'])[:-5]),
                   ('Sender', msg['sender_full_name']),
                   ('Sender\'s Email ID', msg['sender_email'])]),
```
Yes, that sounds reasonable. I have amended it in the latest update.

Hmm, I am a bit wary about using the edit data from messages. Given that the data isn't dynamic and needs to be updated, it feels like an overhead. Moreover, the respective key is likely to be changed to a `boolean` in the future.

Related conversation: https://chat.zulip.org/#narrow/stream/3-backend/topic/edit_history/near/879944.

preetmishra commented:
@neiljp Thanks for the feedback.

In the latest update, I have shelved the extra `get_*` method, simplified the patch that you mentioned and improved the `Date & Time` label behaviour. I have also replied to your other in-line comments.

preetmishra commented:
@neiljp Thanks for the review. :+1: 

In the latest update, I have:
* Incorporated the suggested commit order.
* Separated the `fetch_message_history` test into `*_success` and `*_error`.
* Added a comment re why we just need the message ID in `TestEditHistoryView`.
* Parametrized the edit history keypress test over `index` and `initial_dict`.
* Swapped `tag` with `time` in the `EditHistoryView` to improve consistency between the timestamps.

preetmishra commented:
Update to resolve conflicts.

neiljp commented:
@preetmishra Thanks for persisting with this and building upon the work of @shreyamalviya :+1: I've just merged this after a very minor commit text change :tada:

This is another good feature to have in ZT and with some good subtleties to the UI, and taking into account some slightly unexpected responses from the server :+1:
